pyomo/contrib/mindtpy/tabu_list.py

```diff
@@ -1,7 +1,23 @@
-from cplex.callbacks import IncumbentCallback
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2022
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+#
 
+from pyomo.common.dependencies import attempt_import
 
-class IncumbentCallback_cplex(IncumbentCallback):
+cplex, cplex_available = attempt_import('cplex')
+
+
+class IncumbentCallback_cplex(
+    cplex.callbacks.IncumbentCallback if cplex_available else object
+):
```

Comment:
Could this cause some strange errors?

Reply:
Good catch.  This has been updated to raise `DeferredImportError` on any attempt to either create the class or access any class attribute.

Reply:
It is not obvious to me how you did so. Can you explain?

Reply:
(I'm just curious)

Reply:
The `pyomo.common.dependencies.UnavailableClass(cplex)` is a (new) function that returns a custom base class that wraps the `ModuleUnavailable` object returned by `attempt_import` when the module is not available.  This custom base class implements two methods:
- `__new__` will raise the `DeferredImportError` if the derived class is ever instantiated
- `__getattr__` (on a custom `UnavailableMeta` metaclass) will raise the `DeferredImportError` if anyone attempts to access class / static methods on the derived class (important for getting something other than an `AttributeError` for things like `QtApp.launch_instance()` below)